### PR TITLE
Server

### PR DESCRIPTION
**Pull request by oleander-twig**
I updated schemas in types.json and changed app.py

### CODE REVIEW

**Comment by oleander-twig** [comment]
Created db and tables
Added new queue 
Added queue size as argument 
Changed API

File: app.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+    if not importer.is_valid_repo(repo):
+        return "BAD REQUEST", 400
+    
+    repos_in_db = client.query_row(f"SELECT COUNT(repo_name) FROM git.commits WHERE repo_name = '{repo}'")
+    if repos_in_db[0] >= 0:
+        return "ALREADY_PROCESSED", 200
+    
+    repos_in_queue = client.query_row(f"SELECT COUNT(repo_name) FROM git.new_queue WHERE repo_name = '{repo}'")
+    if repos_in_queue[0] >= 0:
```
Comment (gingerwizard) [inline app.py:47]: if you use `schedule_repo_job` (see below) this check is done for you and you can remove.

File: app.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+        return "ALREADY_PROCESSING", 200
+    
+    max_queue_size = args.size
+
+    queue_size = client.query_row("SELECT COUNT(repo_name) FROM git.new_queue")
+    if queue_size == max_queue_size:
+        return "QUEUE IS FULL", 403
+
+    client.query_row(f"INSERT INTO git.new_queue (repo_name) VALUES ('{repo}')")
```
Comment (gingerwizard) [inline app.py:56]: use `schedule_repo_job` method - task_table param is your queue name. See the exceptions which can be thrown i.e.
1. Queue size exceeded
2. Already scheduled

Note: currently we catch the exceptions in `schedule_repo_job`. May we throw typed exceptions there instead - that you can catch here. I'll then change other code to catch these as well.

File: create_db.py [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+    password=config["password"],
+    secure=config["secure"],
+)
+client = RepoClickHouseClient(clickhouse)
+
+client.query_row("CREATE DATABASE IF NOT EXISTS git")
+
+client.query_row(
+    """
+CREATE TABLE git.commits
```
Comment (gingerwizard) [inline create_db.py:33]: use schema in config

File: create_db.py [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+    `hunks_changed`  UInt32,
+    `repo_name`      LowCardinality(String),
+    `updated_at`     DateTime MATERIALIZED now()
+) ENGINE = ReplacingMergeTree
+ORDER BY (repo_name, time, hash)"""
+)
+
+client.query_row(
+    """
+CREATE TABLE git.file_changes
```
Comment (gingerwizard) [inline create_db.py:56]: see above - schema is in types.json

File: create_db.py [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+    `repo_name`             LowCardinality(String),
+    `updated_at`            DateTime MATERIALIZED now()
+) ENGINE = ReplacingMergeTree
+ORDER BY (repo_name, time, commit_hash, path)
+SETTINGS index_granularity = 8192"""
+)
+
+client.query_row(
+    """
+CREATE TABLE git.line_changes
```
Comment (gingerwizard) [inline create_db.py:89]: again types.json

File: create_db.py [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+    `commit_hunks_changed`       UInt32,
+    `repo_name`                  LowCardinality(String),
+    `updated_at`                 DateTime MATERIALIZED now()
+) ENGINE = ReplacingMergeTree
+ORDER BY (repo_name, time, commit_hash, path, line_number_old, line_number_new)"""
+)
+
+client.query_row(
+    """
+CREATE TABLE git.work_queue
```
Comment (gingerwizard) [inline create_db.py:136]: schema can be here but name is in config

File: create_db.py [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+)
+ENGINE = KeeperMap('git_queue')
+PRIMARY KEY repo_name
+
+"""
+)
+
+client.query_row(
+    """
+CREATE TABLE git.new_queue
```
Comment (gingerwizard) [inline create_db.py:152]: schema can be here but name is in config

File: app.py [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+
+    return "OK", 201
+
+if __name__ == '__main__':
+    app.run(host='0.0.0.0', port=5000)
```
Comment (gingerwizard) [inline app.py:61]: nit: newline

**Review by gingerwizard — CHANGES_REQUESTED** [review]

File: app.py (hunk 2 above)
Comment (oleander-twig) [reply app.py:56]: As I can see, there is no specific check for queue size in method `schedule_repo_job`. Should I add it?

File: create_db.py [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+    native_port=config["native_port"],
+    username=config["username"],
+    password=config["password"],
+    secure=config["secure"],
+)
+client = RepoClickHouseClient(clickhouse)
+
+client.query_row("CREATE DATABASE IF NOT EXISTS git")
+
+client.query_row(types[0][3])
```
Comment (gingerwizard) [inline create_db.py:32]: can we just loop over all 3. Also don't release on [3] for the create statement. Index by name.

File: app.py [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+
+app = Flask(__name__)
+
+
+@app.get("/add_new_repo")
+def process():
+    sqs = boto3.client('sqs')
+    repo = request.args.get("repo")
+    
+    try:
```
Comment (gingerwizard) [inline app.py:46]: we dont use sqs anymore. remove.

File: app.py [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+    secure=config["secure"],
+)
+client = RepoClickHouseClient(clickhouse)
+
+app = Flask(__name__)
+
+
+@app.get("/add_new_repo")
+def process():
+    sqs = boto3.client('sqs')
```
Comment (gingerwizard) [inline app.py:43]: we dont use.

File: repo/schedule.py [hunk 12]
```diff
@@ -19,9 +25,9 @@ def _is_job_scheduled(client: RepoClickHouseClient, task_table, repo_name):
 def schedule_repo_job(client: RepoClickHouseClient, sqs: BaseClient, queue_url: str, task_table: str, repo_name: str,
```
Comment (gingerwizard) [inline repo/schedule.py:25]: In the latest version of this code we check the queue size for you

File: app.py [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+        return "BAD REQUEST", 400
+    
+    repos_in_db = client.query_row(f"SELECT COUNT(repo_name) FROM git.commits WHERE repo_name = '{repo}'")
+    if repos_in_db[0] >= 0:
+        return "ALREADY_PROCESSED", 200
+    
+    
+    max_queue_size = args.size
+
+    queue_size = client.query_row("SELECT COUNT(repo_name) FROM git.new_queue")
```
Comment (gingerwizard) [inline app.py:64]: latest version of schedule_job_queue checks queue size - you can modify to throw typed exception (like you have done for others) and catch

**Review by gingerwizard — CHANGES_REQUESTED** [review]

**Review by gingerwizard — APPROVED** [review]

**Pull request merged**